proxy/upgrade.go

```diff
@@ -0,0 +1,195 @@
+package proxy
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// isUpgradeRequest returns true if and only if there is a "Connection"
+// key with the value "Upgrade" in Headers of the given request.
+func isUpgradeRequest(req *http.Request) bool {
+	for _, h := range req.Header[http.CanonicalHeaderKey("Connection")] {
+		if strings.Contains(strings.ToLower(h), "upgrade") {
+			return true
+		}
+	}
+	return false
+}
+
+// getUpgradeRequest returns the protocol name from the upgrade header
+func getUpgradeRequest(req *http.Request) string {
+	for _, h := range req.Header[http.CanonicalHeaderKey("Connection")] {
+		if strings.Contains(strings.ToLower(h), "upgrade") {
+			return strings.Join(req.Header[h], " ")
+		}
+	}
+	return ""
+}
+
+// UpgradeProxy stores everything needed to make the connection upgrade.
+type upgradeProxy struct {
+	backendAddr  *url.URL
+	reverseProxy *httputil.ReverseProxy
+	insecure     bool
+}
+
+// TODO: add user here
+type auditLog struct {
+	Method   string `json:"method"`
+	Path     string `json:"path"`
+	Query    string `json:"query"`
+	Fragment string `json:"fragment"`
+}
+
+// serveHTTP inspects the request and either proxies an upgraded connection directly,
```

Comment:
does it really inspect the request, or that happens before? maybe changed from a previous version?